### PR TITLE
[PGO] Reland PGO's Counter Reset and File Dumping APIs #76471

### PR DESCRIPTION
**Pull request by qiongsiwu**
https://github.com/llvm/llvm-project/pull/76471 caused buildbot failures on Windows. For more details, see https://github.com/llvm/llvm-project/issues/77546. 

This PR revises the test and relands https://github.com/llvm/llvm-project/pull/76471.

### CODE REVIEW

**Comment by minglotus-6** [comment]
thanks for the fix!

One test `Clang :: Preprocessor/init.c` failed (https://lab.llvm.org/buildbot/#/builders/74/builds/24843) and the error is due to mismatched macros (if I'm reading correctly wants `__LONG_LONG_MAX__` right after `__LLONG_WIDTH__` but sees `__LLVM_INSTR_PROFILE_GENERATE` in the middle). I'm wondering if the test (and other tests like `Preprocessor/init-aarch64.c`) needs update (or if they should be fixed by https://github.com/llvm/llvm-project/issues/77546#issuecomment-1887953628)

**Comment by qiongsiwu** [comment]
> thanks for the fix!
> 
> One test `Clang :: Preprocessor/init.c` failed (https://lab.llvm.org/buildbot/#/builders/74/builds/24843) and the error is due to mismatched macros (if I'm reading correctly wants `__LONG_LONG_MAX__` right after `__LLONG_WIDTH__` but sees `__LLVM_INSTR_PROFILE_GENERATE` in the middle). I'm wondering if the test (and other tests like `Preprocessor/init-aarch64.c`) needs update (or if they should be fixed by [#77546 (comment)](https://github.com/llvm/llvm-project/issues/77546#issuecomment-1887953628))

Thanks for the fast response @minglotus-6 ! The preprocessor `init` tests were failing due to a mis-optimization introduced recently. Due to the misoptimization, `hasProfileInstr()` ([here](https://github.com/llvm/llvm-project/pull/78285/files#diff-3e17619db20856d476f7eee9bd87f730d9594b4aee2b796f6829a55de4c1fbf0R498)) always returned true and hence the `__LLVM_INSTR_PROFILE_GENERATE` macros were always inserted when it should not show up. As noted in the comment, https://github.com/llvm/llvm-project/pull/77831 fixed the mis-optimization. `hasProfileInstr()` is slightly rewritten in this PR as well to avoid triggering the pattern that was mis-optimized. 

That said this PR still has some issues. Four seemingly unrelated tests are failing the pre-commit CI on Linux and I am investigating. I will update this PR once these failures are sorted out! Thanks for your patience! 

```
Failed Tests (4):
  cfi-devirt-lld-thinlto-x86_64 :: cross-dso/icall/dlopen.cpp
  cfi-devirt-lld-x86_64 :: cross-dso/icall/dlopen.cpp
  cfi-standalone-lld-thinlto-x86_64 :: cross-dso/icall/dlopen.cpp
  cfi-standalone-lld-x86_64 :: cross-dso/icall/dlopen.cpp
```

**Comment by qiongsiwu** [comment]
It seems that either the precommit CI has some problems, or there are existing `compiler-rt` breakages. https://github.com/llvm/llvm-project/pull/78037 is also failing the same test cases. The changes in #78037 does not look harmful either. I took a brief look at the current buildbots and I don't see any similar failures. Additionally, I am not able to reproduce these failures locally. 

I will see what can be investigated next, but I suspect at this PR is not the root cause of the `dlopen.cpp` failures.

**Comment by qiongsiwu** [comment]
Ok I think I have a good theory on why these tests are failing. Most likely the test machine does not have the latest `lld` during the test. If we build `lld` together with the project, these tests pass.

File: compiler-rt/test/profile/Inputs/instrprof-api.c.profdata
Comment (w2yehia) [inline compiler-rt/test/profile/Inputs/instrprof-api.c.profdata:1]: will need instructions for how to recreate this file in the future.

**Comment by w2yehia** [comment]
@qiongsiwu explained to me offline that the issue on Windows is that calls __llvm_orderfile_dump fail so he's not able to execute the instrumented program if it calls that function.
Regarding the `compiler-rt/test/profile/instrprof-api.c` test. The disadvantage of having a profdata file is that it's hard to debug/reproduce; the Inputs directory currently has no profdata files, it's all source/text files. So, it's better to avoid storing the profdata.
We can disable the test on Windows.
Or keep the Windows coverage and `#ifdef` the `__llvm_orderfile_dump` API call on Windows (which will require versioning the expected output using LIT substitution magic).

**Comment by qiongsiwu** [comment]
Thanks for to comments/suggestions @w2yehia ! The test is modified so that

1. The profile files are generated. 
2. The test does not support Windows to avoid known limitations on Windows.

**Review by w2yehia — APPROVED** [review]

**Comment by qiongsiwu** [comment]
Hi @vitalybuka ! I am landing this PR. The precommit CI failures do not seem to related to changes in this PR. Please let me know if you see failures and I will revert. Thanks so much!

**Comment by rorth** [comment]
This patch broke both the [Solaris/sparcv9](https://lab.llvm.org/buildbot/#/builders/72/builds/2015) and [Solaris/amd64](https://lab.llvm.org/staging/#/builders/94/builds/394) buildbots.

If it really requires recent `lld`, this is guaranteed to not work: `lld` hasn't even been ported to Solaris.  Similar issues most likely will exist on other targets where the bot either doesn't build `lld` or it isn't supported at all.

**Comment by qiongsiwu** [comment]
> This patch broke both the [Solaris/sparcv9](https://lab.llvm.org/buildbot/#/builders/72/builds/2015) and [Solaris/amd64](https://lab.llvm.org/staging/#/builders/94/builds/394) buildbots.
> 
> If it really requires recent `lld`, this is guaranteed to not work: `lld` hasn't even been ported to Solaris. Similar issues most likely will exist on other targets where the bot either doesn't build `lld` or it isn't supported at all.

Thanks for the note @rorth . I will investigate today. This PR in itself does not require `lld`.

**Comment by qiongsiwu** [comment]
@w2yehia I think a reasonable way to go forward is to remove the testing of `__llvm_orderfile_dump` from the main test, and add a separate test that runs only on Darwin. 

What do you think?

**Comment by qiongsiwu** [comment]
@w2yehia and I discussed, and we will remove `__llvm_orderfile_dump` from the main test for now.

**Comment by qiongsiwu** [comment]
https://github.com/llvm/llvm-project/pull/79150 is posted to avoid calling `_llvm_orderfile_dump()` in the test.

**Pull request merged**